### ISSUE TITLE
reuse_distance and reuse_time trace analyzers need to compute per thread

### ISSUE THREAD

**Issue by derekbruening** [DynamoRIO/dynamorio#3227]
The drcachesim trace analyzers for reuse_distance and reuse_time need to perform their reuse calculations on a per-thread basis, and then aggregate across threads, rather than considering the merged multi-thread memory access stream when counting reuses.

**Comment by derekbruening** [comment]
Tracking per-thread state incurs a large performance hit:

TOT:
```
$ rm -rf drmemtrace.*.dir; bin64/drrun -t drcachesim -offline -- ~/dr/test/threadsig 32 2000000
$ du -sh drmemtrace.threadsig.*
3.7G	drmemtrace.threadsig.87631.8869.dir
$ bin64/drrun -t drcachesim -simulator_type basic_counts -indir drmemtrace.threadsig.*.dir; /usr/bin/time bin64/drrun -t drcachesim -simulator_type reuse_distance -indir drmemtrace.threadsig.*.dir
37.94user 0.05system 0:38.02elapsed 99%CPU (0avgtext+0avgdata 8564maxresident)k
$ l drmemtrace.threadsig.*.dir/trace
total 55M
1.2M drmemtrace.threadsig.87631.3027.trace.gz  1.8M drmemtrace.threadsig.87642.0813.trace.gz  1.7M drmemtrace.threadsig.87653.0160.trace.gz
1.8M drmemtrace.threadsig.87632.4087.trace.gz  1.8M drmemtrace.threadsig.87643.6941.trace.gz  1.7M drmemtrace.threadsig.87654.2804.trace.gz
```

Per-thread, serial:
```
$ /usr/bin/time bin64/drrun -t drcachesim -simulator_type reuse_distance -indir ../build_x64_rel_tests/drmemtrace.threadsig.*.dir -jobs 0
282.69user 0.01system 4:42.71elapsed 99%CPU (0avgtext+0avgdata 10560maxresident)k
```

That's a 7.4x slowdown.

Leveraging the parallelism from #3230 we have:

Per-thread, parallel:
```
$ /usr/bin/time bin64/drrun -t drcachesim -simulator_type reuse_distance -indir ../build_x64_rel_tests/drmemtrace.threadsig.*.dir
307.42user 0.15system 0:50.77elapsed 605%CPU (0avgtext+0avgdata 10956maxresident)k
```

That's close to what it was, but it's still 1.3x slower.

**Comment by derekbruening** [comment]
This should have been closed by PR #3345 f9dc5f5 but it accidently listed #3345!
￼
